### ISSUE TITLE
Can not sign large transactions - URL too long

### ISSUE THREAD

**Issue by ilblackdragon** [near/near-wallet#1822]
If transaction is relatively large, transaction info doesn't fit into URL as there is only 2048 characters in URL can be.
Passing transaction data via POST should be supported.

**Comment by stefanopepe** [comment]
will have @MaximusHaximus look at it, as it seems a feature related to the wallet helper to me

**Comment by MaximusHaximus** [comment]
@ilblackdragon Currently, `wallet.near.org` is a static site that cannot be POSTed to; any routes that terminate at `wallet.near.org` are GET only, and are always passed along to the client side router to be handled. There is no NEAR-owned web server backing this static site; serving static resources is provided by `netlify` and `render`.

**Brainstorming, I think there are 3 primary ways that we can handle this case:**

1. **Convert `wallet.near.org` from a static site to a full-fledged dynamic web server instance that we can control, much like our `near-contract-helper` backend.  We could then handle both GET and POST requests.**
- PRO: Backend routes could use the same hostname as client-side routes; e.g. `near-contract-helper` routes could become routes of `wallet.near.org` instead of existing on a different host/URL.
- CON: This would require that we manage the resources to handle simple web requests to load our static site and static resources; right now this is handled for us by Netlify and Render.com.  Ideally, we run multiple web instance nodes to handle spikes in usage, behind a load balancer to ensure reliability for serving up the static site and providing POST routes on `wallet.near.org`.
- CON: The simple, static sites that are deployed automatically for branch testing and staging testing would need to be replaced with code that would build and deploy instances of a full-fledged web server so that the staging sites can accept POST requests also.
- CON: We would need to modify the deployment process for for both `netlify` and `render` since we use a different hosting service for testnet wallet vs. mainnet wallet, **or** standardize both testnet and mainnet wallet hosting to use the same hosting provider (netlify is preferred, AFAIK)

Ideally, we'd like to keep the main static wallet site as an entirely static site if possible for simplicity.

2. **Involve `near-contract-helper` in the transaction signing process.**
- We could POST a transaction to a route that `near-contract-helper` provides, and return a unique identifier for that transaction back to the browser.  The browser can then redirect the user to the wallet frontend, and instead of providing the entire transaction in the URL, it could _just_ provide the unique transaction ID that was returned from the POST to the `near-contract-helper` backend.  The wallet UI can then load that transaction via the `near-contract-helper`, sign and send it, and then remove the transaction from the `near-contract-helper` backend.
- PRO: Serving static resources would still be handled by netlify and render.com's simple static site hosting infrastructure.
- CON: Submitting transactions to the wallet frontend would require the `near-contract-helper` to be involved; as it is currently designed, only the static site needs to be loadable for simple transactions.
- MITIGATION: We could only use the `near-contract-helper` as an interim to store transactions over a specific size/limit of characters, so simple/short transactions could still be sent directly via URL parameters to the wallet frontend, but larger transactions would need to be placed into an interim store that they could be loaded from by the wallet frontend.

2a. **ALTERNATIVELY:  We wouldn't _necessarily_ have to store the transactions on the `near-contract-helper` backend. In theory, the large transaction could be hosted _anywhere_ that the wallet users' browser can reach, and the URL where that transaction data exists could be passed to the wallet frontend UI in the query string instead of a unique hash identifying the transaction.**
- The wallet UI would load, detect the presence of a URL parameter that contains an URL that points to a large transaction that the user wants to sign & submit, and fetch the transaction content from that URL using an XHR GET request.  
- PRO: We wouldn't need the contract helper to be involved in passing the large transaction to the wallet; any public hosting could be used (IPFS? S3?)
- CON: The content of the transaction would need to be put somewhere that the wallet user's browser could fetch it from, which would assumedly also be accessible to anyone else who can view that public site, and might not be secure depending on the hosting service that the URL loads from. It feels safer to me for us to hold on to the transaction for the user on the `near-contract-helper` backend rather than allow the frontend wallet to fetch transactions from 'random' URLs that could be unsafe/compromised.  With an opaque UUID and a route on `near-contract-helper`, we could be confident that the large transaction content that the wallet loads is exactly what it had submitted.

3. Use a pattern similar to how our 2FA implementation works, where the client submits transactions to a contract on-chain directly, for execution later when 2FA is verified.  This allows the client to directly submit a large transaction, then the wallet UI transaction could be a simpler, smaller transaction that allows the user to approve the transaction that they had already signed and submitted.
- I remember asking Vlad about this during my onboarding -- From what I understand, there is a limit to the size of a transaction that Ledger devices can sign which may make this pattern a requirement for ledger users even if we had POST abilities to send larger payloads, because the larger transactions would fail to be signed on the Ledger device.
- @vgrichina do you remember what the size limit is for signing NEAR transactions on ledger devices?

cc: @stefanopepe

**Comment by starpause** [comment]
@MaximusHaximus "Use a pattern similar to how our 2FA implementation works" sounds the best to me. It's a tested approach and reinforces a convention for handling such circumstances. Does it have CONS?

**Comment by MaximusHaximus** [comment]
> @MaximusHaximus "Use a pattern similar to how our 2FA implementation works" sounds the best to me. It's a tested approach and reinforces a convention for handling such circumstances. Does it have CONS?

The main con there (and it's a big one), is that it requires that a smart contract exists on the account, where the transactions can be stored in that smart-contract on-chain in between them being submitted and approved... which costs storage and means your NEAR account must have that smart contract deployed to it.

References:
https://github.com/near/core-contracts/tree/master/multisig
https://github.com/near/core-contracts/blob/master/multisig/src/lib.rs#L116

**Comment by vgrichina** [comment]
Ledger and 2FA have even stricter limitation on transaction size. URL actually can be much longer than 2048 chars in most browsers, dunno where you take this number @ilblackdragon 

@MaximusHaximus Ledger has less than 1k limit for transaction size, as it has 2 or 4K RAM total. 

So to solve this in a way that works for all cases (including Ledger) it should be solved on app side:
1) send large transaction using app specific key (this just creates request)
2) confirm the submitted request with a smaller transaction submitted through wallet. This smaller transaction content can be something user can verify for real (e.g. smart contract code hash if the request was to deploy smart contract)

so basically similar to what is in @MaximusHaximus last variant.

**Comment by MaximusHaximus** [comment]
> Ledger and 2FA have even stricter limitation on transaction size. URL actually can be much longer than 2048 chars in most browsers, dunno where you take this number @ilblackdragon
> 
> @MaximusHaximus Ledger has less than 1k limit for transaction size, as it has 2 or 4K RAM total.
> 
> So to solve this in a way that works for all cases (including Ledger) it should be solved on app side:
> 
> 1. send large transaction using app specific key (this just creates request)
> 2. confirm the submitted request with a smaller transaction submitted through wallet. This smaller transaction content can be something user can verify for real (e.g. smart contract code hash if the request was to deploy smart contract)
> 
> so basically similar to what is in @MaximusHaximus last variant.

1. The 2048 char issue is caused by infrastructure between the user and the browser; a lot of older proxies and webservers still have the 2048char limit on URLs, even though modern e.g. Chrome etc. will allow huge URLs, So it is still best practice to keep URL length under 2048 chars.

2. Re: the Ledger limitation, it would appear that there is a way to sign large transactions with the Ledger, by streaming the transaction data to be signed in chunks to the device, rather than trying to send the entire transaction data for signing as a single interaction with the device (thanks @SkidanovAlex for the insights!).

3. Re: the URL size limitation, @evgenykuzyakov suggested we look at https://walletconnect.org/ to resolve this problem.  I'll be checking that out as a solution -- if anyone has any objections re: the https://walletconnect.org/ protocol I'd love to hear them here. cc: @stefanopepe @vgrichina

**Comment by abacabadabacaba** [comment]
Just noting that [WalletConnect](https://walletconnect.org/) is an option that doesn't have the disadvantages of the first three. It can be used from a static website and doesn't post anything extra on the chain. It requires a relayer, but that relayer doesn't need to be NEAR-specific, in fact, existing WalletConnect relayers should work.

**Comment by vgrichina** [comment]
> Re: the Ledger limitation, it would appear that there is a way to sign large transactions with the Ledger, by streaming the transaction data to be signed in chunks to the device, rather than trying to send the entire transaction data for signing as a single interaction with the device, so this restriction appears to be more specifically due to how the NEAR Ledger app works with the ledger device than an outright constraint of the ledger itself. (thanks @SkidanovAlex for the insights!).

of course, you might stream large transaction into Ledger. I am well aware of this possibility. However:
1) It is complicated to implement, making it easy to screw up security of implementation.
2) There is no reasonable way for user to verify such transaction on a tiny Ledger screen, making the whole process of signing with Ledger more or less pointless (as attacker can make you sign whatever stream of data necessary).

> There is no reasonable way for user to verify such transaction on a tiny Ledger screen

Note that this is a serious security UX issue more or less relevant for any of the proposed way for user to actually sign large transaction (vs avoid having user to confirm large transaction). The only things user can reasonably verify in this case gonna be:
- sender
- receiver
- attached NEAR tokens.

**Comment by vgrichina** [comment]
> Re: the URL size limitation, @evgenykuzyakov suggested we look at https://walletconnect.org/ to resolve this problem. I'll be checking that out as a solution -- if anyone has any objections re: the https://walletconnect.org/ protocol I'd love to hear them here. cc: @stefanopepe @vgrichina

last time I checked experience using it with Rainbow Wallet was pretty suboptimal:
- stuff is glitchy and causing delays
- to use Rainbow Wallet with mobile website you'd have to either have 2 devices to use camera to scan QR, or try to workaround it with screenshots.

I dunno whether these are limitations of WalletConnect in general or Rainbow specifically.

In any case I think the main concern is security of given UX.  If you let user sign transaction with more than 2k worth of data they almost certainly cannot verify what inside.

What is the point of making it operation that is confirmed by user, when user cannot do any reasonable verification?

**Comment by vgrichina** [comment]
> * The 2048 char issue is caused by infrastructure between the wallet host and the browser; a lot of older proxies and webservers still have the 2048char limit on URLs, even though modern e.g. Chrome etc. will allow huge URLs, So it is still best practice to keep URL length under 2048 chars.
> *

what some older proxies or webservers do is irrelevant because we can chose our backend infra. In any case you can avoid server-side limitation by passing relevant part in URL fragment (i.e. after `#`).

**Comment by MaximusHaximus** [comment]
> > * The 2048 char issue is caused by infrastructure between the wallet host and the browser; a lot of older proxies and webservers still have the 2048char limit on URLs, even though modern e.g. Chrome etc. will allow huge URLs, So it is still best practice to keep URL length under 2048 chars.
> 
> what some older proxies or webservers do is irrelevant because we can chose our backend infra. In any case you can avoid server-side limitation by passing relevant part in URL fragment (i.e. after `#`).

Sadly, we don't control what hardware our requests have to traverse to get to the web clients, even if we _can_ control our hosting infrastructure, so this isn't irrelevant.

Furthermore, IE11 (still in support through 2022) has a limit of 2083 characters and even versions of Edge published through mid 2020 only supported up to around 4000 characters in a URL, and older versions of edge still had the same limit as IE @ 2083 characters.

As long as we want the wallet to be accessible for a wide range of browsers and consumer networks, this limit is still relevant.

>Note that this is a serious security UX issue more or less relevant for any of the proposed way for user to actually sign large transaction (vs avoid having user to confirm large transaction). The only things user can reasonably verify in this case gonna be:
> sender
> receiver
> attached NEAR tokens.

I actually think that depending on the data, 2kb is sometimes not too much for someone to read + verify...for example, when a payload may have e.g. large unicode characters that are 3 or 4 bytes each -- this could be only a few hundred characters in those cases.

I'm having a hard time modeling what it would look like to "avoid having user to confirm large transaction"; is this _always_ possible for all use cases?  Is it not reasonable that sometimes people just need to sign a large transaction?

I'd really appreciate an example of how we can accomplish the goal of allowing everyone to do all of the things they need in transactions that are <2kb in size, with ledger signing in a generic way that still provides the kind of manual verification ability that you are describing.  Even a brief description of how we would accomplish this with a 'real app example' would be enlightening

**Comment by vgrichina** [comment]
> I'd really appreciate an example of how we can accomplish the goal of allowing everyone to do all of the things they need in transactions that are <2kb in size, with ledger signing in a generic way that still provides the kind of manual verification ability that you are describing. Even a brief description of how we would accomplish this with a 'real app example' would be enlightening

**Comment by vgrichina** [comment]
> Sadly, we don't control what hardware our requests have to traverse to get to the web clients, even if we _can_ control our hosting infrastructure, so this isn't irrelevant.

wallet website uses HTTPS, so generally there shouldn't be proxies in the middle (in normal situation). If there are some – we have a larger problem than URL length.

> 
> I'm having a hard time modeling what it would look like to "avoid having user to confirm large transaction"; is this _always_ possible for all use cases? Is it not reasonable that sometimes people just need to sign a large transaction?
> 

Every operation which is large transaction requiring user approval can be split into 2 transactions:
1) is confirmed by user (and so small) and requires full access key 
2) is done using app specific key (and so can be max possible size)

I would be interested to see examples which you think cannot be split into 2 transactions. I guess it's gonna be the same stuff that is impossible to do with 2FA (AFAIK there is only limitation with having less gas).

> I'd really appreciate an example of how we can accomplish the goal of allowing everyone to do all of the things they need in transactions that are <2kb in size, with ledger signing in a generic way that still provides the kind of manual verification ability that you are describing. Even a brief description of how we would accomplish this with a 'real app example' would be enlightening

E.g. upgrade a smart contract:
1) User sends transaction using full access key to confirm which contract hash needs to be deployed and attach NEAR for it's storage. Contract hash gets confirmed by user and recorded. This transaction also can add app access key for next transaction if needed.
2) Whoever has app access key submits full binary of new smart contract to deploy. This binary is hashed and verified against hash recorded on chain. If valid – new version is deployed. App access key can be removed at this point.
